File: pipeline/dp_cpp/src/streaming.cpp
#include "streaming.hpp"
#include "common.hpp"

#include <algorithm>
#include <cmath>
#include <queue>
#include <stdexcept>
#include <unordered_map>
#include <unordered_set>
#include <utility>

namespace {

// Min-heap ordering on the quantised (t, d) key — exact lexicographic order
// on integers, no float-tie hazards.
struct KeyGreater {
    bool operator()(const TDKey& a, const TDKey& b) const noexcept {
        if (a.t9 != b.t9) return a.t9 > b.t9;
        return a.d9 > b.d9;
    }
};

}  // namespace

StreamingResult solve_streaming(const Frame& frame, double eta,
                                const TimeKey& time_key,
                                double d_start,
                                bool node_first) {
    if (!node_first)
        throw std::runtime_error(
            "streaming engine supports node-first only; use the legacy "
            "engine for the speed-first grid");

    const double L = frame.cfg.length_nm;
    const TDKey src = make_td_key(0.0, d_start);

    std::unordered_map<TDKey, double> cost;
    std::unordered_map<TDKey, AtomicEdge> parent;  // winning incoming arc
    // exact coordinates as first discovered (node-first destinations are
    // generated already 9-rounded, so these coincide with the key values;
    // kept anyway so the enumerator receives the same doubles the legacy
    // builder interned).
    std::unordered_map<TDKey, std::pair<double, double>> coord;
    std::unordered_set<TDKey> popped;

    std::priority_queue<TDKey, std::vector<TDKey>, KeyGreater> heap;
    cost.emplace(src, 0.0);
    coord.emplace(src, std::make_pair(0.0, d_start));
    heap.push(src);

    StreamingResult out;

    while (!heap.empty()) {
        TDKey k = heap.top();
        heap.pop();
        if (!popped.insert(k).second) continue;

        const auto [t, d] = coord.at(k);
        const double base = cost.at(k);

        // Same call the legacy builder makes (forecast_hour = -1,
        // override_sample_hour = -1 → time-varying actual weather with
        // NaN walkback), node-first enumeration.
        auto edges = emit_from_src(t, d, frame, /*forecast_hour=*/-1,
                                   /*override_sample_hour=*/-1, time_key,
                                   /*node_first=*/true);
        for (auto& e : edges) {
            ++out.n_edges_evaluated;
            if (std::isnan(e.fuel_mt)) continue;  // legacy-solver parity
            TDKey dk = make_td_key(e.dst_t, e.dst_d);
            if (popped.count(dk))
                throw std::runtime_error(
                    "streaming invariant violated: relax into a popped state "
                    "(pop order not topological)");
            const double new_cost = base + e.fuel_mt;
            auto it = cost.find(dk);
            const bool fresh = (it == cost.end());
            if (fresh || new_cost < it->second) {
                if (fresh) {
                    cost.emplace(dk, new_cost);
                    coord.emplace(dk, std::make_pair(e.dst_t, e.dst_d));
                } else {
                    it->second = new_cost;
                }
                parent.insert_or_assign(dk, e);
            }
            if (fresh) heap.push(dk);
        }
    }

    out.n_states = cost.size();

    // ---- sink selection (mirrors BellmanSolver::best_sink, hard ETA) ----
    bool found = false;
    TDKey best{};
    double best_cost = 0.0;
    for (const auto& [k, c] : cost) {
        const auto& [t, d] = coord.at(k);
        if (std::abs(d - L) >= 1e-9) continue;
        if (t > eta + 1e-6) continue;
        if (!found || c < best_cost) {
            found = true;
            best = k;
            best_cost = c;
        }
    }
    if (!found) {
        // distinguish the two legacy error messages
        bool any_sink = std::any_of(cost.begin(), cost.end(), [&](const auto& kv) {
            return std::abs(coord.at(kv.first).second - L) < 1e-9;
        });
        throw std::runtime_error(any_sink ? "No sink reachable within ETA"
                                          : "No sink reachable from source.");
    }

    // ---- backtrack (mirrors BellmanSolver::backtrack) ----
    std::vector<AtomicEdge> path;
    TDKey cur = best;
    while (true) {
        auto it = parent.find(cur);
        if (it == parent.end()) break;
        const AtomicEdge& e = it->second;
        path.push_back(e);
        cur = make_td_key(e.src_t, e.src_d);
        if (cur == src) break;
    }
    std::reverse(path.begin(), path.end());

    out.total_fuel_mt = best_cost;
    out.voyage_time_h = coord.at(best).first;
    out.path = std::move(path);
    return out;
}

File: pipeline/dp_rebuild/streaming.py
"""
Streaming Bellman engine — Phase 2 of the refactor
(docs/refactor_streaming_design.md, approved 2026-08-06).

One forward pass does discovery, discretisation, pricing AND valuation:
states are popped from a lexicographic-(t, d) min-heap; each popped state's
out-edges are produced by the SAME per-source enumerator the legacy builder
uses (`atomic_edges._emit_from_src` = A(d,t) candidates + arc_cost pricing)
and relaxed on the spot. Nothing is stored beyond the two records the answer
needs per state: its cost and its winning parent arc. The arc set is never
materialised.

Bit-exactness vs the legacy build->solve pipeline, by construction:
  - every arc strictly increases t, so heap pops come out in globally
    sorted lex(t, d) order — exactly the topological order the legacy
    solver iterates;
  - per-source edge order is identical (same enumerator);
  - hence the global relaxation sequence, every float operation, the
    parent choices (strict <, first-wins) and the backtracked schedule
    are identical.
  - node-first destinations are generated already rounded to the key
    precision, so rounded keys coincide with raw coordinates.

Scope: node-first only (the speed-first grid stays on the legacy engine).
Luo stays legacy (approved decision).
"""
from __future__ import annotations

import heapq
import sys
from math import inf
from pathlib import Path
from typing import Dict, List, Optional, Set, Tuple

_HERE = Path(__file__).resolve().parent
if str(_HERE) not in sys.path:
    sys.path.insert(0, str(_HERE))

from atomic_edges import AtomicEdge, _emit_from_src
from bellman import BellmanResult
from frame import Frame

_KEY_PRECISION = 9


def _key(t: float, d: float) -> Tuple[float, float]:
    return (round(t, _KEY_PRECISION), round(d, _KEY_PRECISION))


class StreamingStats:
    __slots__ = ("n_nodes", "n_edges_evaluated")

    def __init__(self) -> None:
        self.n_nodes = 0
        self.n_edges_evaluated = 0


def solve_streaming(
    frame: Frame,
    *,
    eta: float,
    eta_mode: str = "hard",
    lam: Optional[float] = None,
    time_key=None,
    d_start: float = 0.0,
    node_first: bool = True,
    forecast_hour: Optional[int] = None,
    override_sample_hour: Optional[int] = None,
    perturber=None,
    wait_mode: str = "off",
) -> Tuple[BellmanResult, StreamingStats]:
    """Solve the DP in a single streaming pass. Returns (result, stats);
    ``stats.n_edges_evaluated`` must equal the legacy ``len(edges)`` — the
    regression harness gates on it."""
    if not node_first:
        raise NotImplementedError(
            "streaming engine supports node-first only; use --engine legacy "
            "for the speed-first grid")

    L = frame.cfg.length_nm
    src_key = _key(0.0, d_start)

    cost: Dict[Tuple[float, float], float] = {src_key: 0.0}
    parent: Dict[Tuple[float, float], Optional[AtomicEdge]] = {src_key: None}
    popped: Set[Tuple[float, float]] = set()
    in_heap: Set[Tuple[float, float]] = {src_key}
    heap: List[Tuple[float, float]] = [src_key]
    stats = StreamingStats()

    while heap:
        k = heapq.heappop(heap)
        in_heap.discard(k)
        if k in popped:
            continue
        popped.add(k)
        t, d = k
        base = cost[k]
        for e in _emit_from_src(t, d, frame,
                                forecast_hour=forecast_hour,
                                override_sample_hour=override_sample_hour,
                                perturber=perturber,
                                time_key=time_key,
                                node_first=True,
                                wait_mode=wait_mode):
            stats.n_edges_evaluated += 1
            dk = _key(e.dst_t, e.dst_d)
            if dk in popped:
                # impossible while every arc strictly increases t — if it
                # ever fires, the pop order was not topological.
                raise AssertionError(
                    f"streaming invariant violated: relax into popped state "
                    f"{dk} from {k}")
            new_cost = base + e.fuel_mt
            prev = cost.get(dk)
            if prev is None or new_cost < prev:
                cost[dk] = new_cost
                parent[dk] = e
            if dk not in in_heap and dk not in popped and prev is None:
                heapq.heappush(heap, dk)
                in_heap.add(dk)

    stats.n_nodes = len(cost)

    # --- sink selection (mirrors BellmanSolver.best_sink) -----------------
    sinks = [k for k in cost if abs(k[1] - L) < 1e-9]
    if not sinks:
        raise ValueError("No sink reachable from the source.")
    if eta_mode == "hard":
        in_time = [k for k in sinks if k[0] <= eta + 1e-6]
        if not in_time:
            raise ValueError(
                f"No sink reachable within ETA {eta} h "
                f"(earliest reachable: {min(k[0] for k in sinks):.3f} h)")
        best = min(in_time, key=lambda k: cost[k])
    elif eta_mode == "soft":
        if lam is None:
            raise ValueError("soft ETA requires `lam`")
        best = min(sinks, key=lambda k: cost[k] + lam * max(0.0, k[0] - eta))
    else:
        raise ValueError(f"Unknown eta_mode {eta_mode!r}; use 'hard' or 'soft'.")

    # --- backtrack (mirrors BellmanSolver.backtrack) ----------------------
    path: List[AtomicEdge] = []
    cur = best
    while parent[cur] is not None:
        e = parent[cur]
        path.append(e)
        cur = _key(e.src_t, e.src_d)
        if cur == src_key:
            break
    path.reverse()

    res = BellmanResult(
        total_fuel_mt=cost[best],
        voyage_time_h=best[0],
        schedule=path,
        sink_node=best,
        eta_mode=eta_mode,
        lam=lam,
        nan_edges_skipped=0,
        nodes_reached=len(cost),
        nodes_unreached=0,
    )
    return res, stats
